File: day5/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <array>
#include <regex>
#include <unordered_map>
#include <unordered_set>
#include <sstream>
#include <set>

using namespace std;

class Graph {
    public:
    Graph(set<pair<int, int>> rules, vector<int> &vertices) : m_vertices(vertices.begin(), vertices.end()) {
        for (auto rule : rules) {
            if (m_vertices.find(rule.first) != m_vertices.end() && m_vertices.find(rule.second) != m_vertices.end()) {
                m_graph[rule.first].push_back(rule.second);
            }
        }
        for (int vertex : vertices) {
            m_visited[vertex] = false;
        }
        find_topo_order();
        vector<int> topo_order;
        while (!m_topo_order.empty()) {
            topo_order.push_back(m_topo_order.top());
            m_topo_order.pop();
        }
        m_middle_vertex = topo_order[topo_order.size() / 2];
    }
    int get_middle_vertex() {
        return m_middle_vertex;
    }
    private:
    void find_topo_order() {
        for (int vertex : m_vertices) {
            if (!m_visited[vertex]) {
                dfs(vertex);
            }
        }   
    }
    void dfs(int vertex) {
        m_visited[vertex] = true;
        for (int neighbor : m_graph[vertex]) {
            if (!m_visited[neighbor]) {
                dfs(neighbor);
            }
        }
        m_topo_order.push(vertex);
    }
    unordered_map<int, vector<int>> m_graph;
    unordered_set<int> m_vertices;
    unordered_map<int, bool> m_visited;
    stack<int> m_topo_order;
    int m_middle_vertex;
};

int main() {
    set<pair<int, int>> rules;
    string line;
    uint64_t res_a = 0;
    uint64_t res_b = 0;
    while (getline(cin, line)) {
        if (line.empty()) {
            break;
        }
        regex re(R"((\d+)\|(\d+))");
        smatch match;
        if (regex_match(line, match, re)) {
            rules.insert(make_pair(stoi(match[1].str()), stoi(match[2].str())));
        }
    }
    while (getline(cin, line)) {
        stringstream ss(line);
        int a;
        vector<int> order;
        while (ss >> a) {
            order.push_back(a);
            ss.ignore(1);
        }
        bool valid = true;
        for (int i = 0; i < order.size(); i++) {
            for (int j = i + 1; j < order.size(); j++) {
                if (rules.find(make_pair(order[j], order[i])) != rules.end()) {
                    valid = false;
                    break;
                }
            }
        }
        if (valid) {
            res_a += order[order.size() / 2];
        } else {
            Graph graph(rules, order);
            res_b += graph.get_middle_vertex();
        }
    }
    cout << res_a << endl;
    cout << res_b << endl;
}

File: day5/main.rs
use std::io::{self, BufRead};
use regex::Regex;
use std::collections::{HashSet, HashMap};

struct Graph {
    graph : HashMap<u32, Vec<u32>>,
    vertices : HashSet<u32>,
}

impl Graph {
    fn new(rules : &HashSet<(u32, u32)>, vertices : &Vec<u32>) -> Self {
        let mut graph = HashMap::new(); 

        let vertices_set : HashSet<u32> = vertices.iter().cloned().collect();
        for &(from, to) in rules {
            if vertices_set.contains(&from) && vertices_set.contains(&to) {
                graph.entry(from).or_insert_with(Vec::new).push(to);
            }
        }

        Graph{
            graph : graph,
            vertices : vertices_set,
        }
    }

    fn find_topo_order(&self) -> u32 {
        let mut visited = HashSet::new();
        let mut topo_order= Vec::new();

        for &vertex in &self.vertices {
            if !visited.contains(&vertex) {
                self.dfs(vertex, &mut visited, &mut topo_order);
            }
        }
        topo_order.reverse();
        topo_order[topo_order.len() / 2]
    }

    fn dfs(&self, vertex : u32, visited : &mut HashSet<u32>, topo_order : &mut Vec<u32>) {
        visited.insert(vertex);
        if let Some(neighbors) = self.graph.get(&vertex) {
            for &neighbor in neighbors {
                if !visited.contains(&neighbor) {
                    self.dfs(neighbor, visited, topo_order);
                }
            }
        }
        topo_order.push(vertex);
    }
}

fn main() {
    let mut res_a = 0;
    let mut res_b = 0;
    let mut rules = HashSet::<(u32, u32)>::new();
    let reg_rules = Regex::new(r"^(\d+)\|(\d+)$").unwrap();
    
    // Read rules
    for line in io::stdin().lock().lines() {
        let line = line.unwrap();
        if line.is_empty() {
            break;
        }
        let caps = reg_rules.captures(&line).unwrap();
        let a: u32 = caps[1].parse().unwrap();
        let b: u32 = caps[2].parse().unwrap();
        rules.insert((a, b));
    }
    
    // Process sequences
    for line in io::stdin().lock().lines() {
        let line = line.unwrap();
        let order: Vec<u32> = line.split(',')
            .map(|x| x.parse::<u32>().unwrap())
            .collect();
        
        let mut is_valid = true;
        for i in 0..order.len() {
            for j in (i + 1)..order.len() {
                if rules.contains(&(order[j], order[i])) {
                    is_valid = false;
                    break;
                }
            }
            if !is_valid {
                break;
            }
        }
        
        if is_valid {
            res_a += order[order.len() / 2];
        } else {
            let graph = Graph::new(&rules, &order);
            res_b += graph.find_topo_order();
        }
    }

    println!("{}", res_a);
    println!("{}", res_b);
}